--- sympy-sympy-1.14.0/pyproject.toml.orig	2025-05-22 18:32:08.558630159 -0400
+++ sympy-sympy-1.14.0/pyproject.toml	2025-05-22 18:32:20.255865020 -0400
@@ -127,7 +127,6 @@
     "lxml.*",
     "matchpy.*",
     "matplotlib.*",
-    "mpmath.*",
     "numpy.*",
     "PIL.*",
     "pycosat.*",
--- sympy-sympy-1.14.0/setup.py.orig	2025-05-22 19:57:24.284097279 -0400
+++ sympy-sympy-1.14.0/setup.py	2025-05-22 19:57:53.038593913 -0400
@@ -320,9 +320,9 @@
               'Source': 'https://github.com/sympy/sympy',
           },
           # Set upper bound when making the release branch.
-          install_requires=[
-              'mpmath >= 1.1.0, < 1.4',
-          ],
+#          install_requires=[
+#              'mpmath >= 1.1.0, < 1.4',
+#          ],
           py_modules=['isympy'],
           packages=['sympy'] + modules + tests,
           ext_modules=[],
