diff --git a/cmake/onnxruntime_mlas.cmake b/cmake/onnxruntime_mlas.cmake
index e2c297c0c..c4b722e60 100644
--- a/cmake/onnxruntime_mlas.cmake
+++ b/cmake/onnxruntime_mlas.cmake
@@ -325,7 +325,8 @@ else()
           set_source_files_properties(${MLAS_SRC_DIR}/power/QuantizePowerVSX.cpp PROPERTIES COMPILE_FLAGS "-mcpu=power9")
         endif()

-        check_cxx_compiler_flag("-mcpu=power10" HAS_POWER10)
+        #check_cxx_compiler_flag("-mcpu=power10" HAS_POWER10)
+        set(HAS_POWER10 OFF)
         if(HAS_POWER10)
           set(CMAKE_REQUIRED_FLAGS "-mcpu=power10")
           check_cxx_source_compiles("
